### ISSUE TITLE
RFC: Use `np.int64` by default for CSR matrices' `indices` and `indptr`

### ISSUE THREAD

**Issue by jjerphan** [scipy/scipy#16774]
### Context

When creating a [`scipy.sparse.csr_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html) (for instance using a numpy array), the underlying `indices` and `indptr` arrays either use `np.int32` or `np.int64` depending on `nnz`, the number of non zeros elements (namely `int64` are used when `nnz >= 2**31`):

https://github.com/scipy/scipy/blob/935d537feeb5ff2940f50de75bda01dd21350cb9/scipy/sparse/_sputils.py#L134-L185

This mechanism has been introduced in https://github.com/scipy/scipy/pull/442 to support `np.int64` while conserving the previous implementation which was using `np.int32` when possible based on discussions among maintainers. It has latter been adapted and propagated in sparse matrices' creation via other contributions, notably https://github.com/scipy/scipy/pull/3468, https://github.com/scipy/scipy/pull/4678/.

### Problem

Unfortunately, some algorithms which directly work on CSR arrays necessitate typing `indices` and `indptr`. This is for instance the case of some algorithms implemented in Cython in scikit-learn.

Supporting both `np.int32` and `np.int64` comes with complex code adaptation (and potentially performance regressions) that downstream libraries might not want to maintain (this is the case of scikit-learn for instance).

Moreover, this restricts some algorithms working on sparse matrices in SciPy (namely [`scipy.sparse.csgraph`](https://docs.scipy.org/doc/scipy/reference/sparse.csgraph.html)'s, implemented in Cython) to only work on small problems.

Ideally, the dtype of `indices` and `indptr` could be set at CSR matrices' creation, but this is not controllable by downstream libraries. The only possible control is to cast `indices` and `indptr` to use the dtype chosen by the downstream library and hence create a copy of them. 

### Proposed solution

The creation of SciPy CSR matrices could be changed to use `int64` by default for `indices` and `indptr` while still being able to specify using `int32` if needed. This would allow avoiding to copy and have a better range representation and the cost of a slightly larger memory footprint for CSR matrices for with `nnz < 2**31`.

Also `np.uint`s could be used over `np.uint` because indices and their pointers are positive numbers. This would allow saving memory (as unsigned integers takes twice as less memory than integers) but would also corrupt serialized sparse matrices.

Past discussions in https://github.com/scipy/scipy/pull/442, https://github.com/scipy/scipy/pull/3468 and https://github.com/scipy/scipy/pull/4678/ somewhat mentioned part of the first solution, but I think making this choice might break someone's else usage of sparse matrices.

Before doing any UX and adaptation work, I think it is worth imagining potential scenarios beyond SciPy (and scikit-learn) where this solution is inadequate and I would welcome @pv, @wnbell and @rgommers's experience and expertise on sparse matrices (not required, not a urgency).

I am willing to work on a solution.

### References

Relevant issues and discussions linked to this issue not exhaustively include:

 - [[Scientific python Discourse] Scientific python and sparse arrays (scipy summary + future directions)](https://discuss.scientific-python.org/t/scientific-python-and-sparse-arrays-scipy-summary-future-directions/368/10)
 - https://github.com/scikit-learn/scikit-learn/issues/23653
 - https://github.com/scipy/scipy/issues/16569

**Comment by ogrisel** [comment]
> The creation of SciPy CSR matrices could be changed to use int64 by default for indices and indptr while still being able to specify using int32 if needed. This would allow avoiding to copy and have a better range representation and the cost of a slightly larger memory footprint for CSR matrices for with nnz < 2**31.

Using `int64` by default, even when `nnz < np.iinfo(np.int32).max` seems like a performance regression (from a memory usage point of view to me).

I think it would be convenient to make to possible to pass an index dtype explicitly at sparse matrix creation time but I don't think we should change the default "automatic" logic based on actual contents.

**Comment by ivirshup** [comment]
> I think it would be convenient to make to possible to pass an index dtype explicitly at sparse matrix creation time but I don't think we should change the default "automatic" logic based on actual contents.

I sorta agree with this. I think the main work to be done is removing steps which automatically downcast the array. For example:

```python
>>> type(X), X.indptr.dtype
(scipy.sparse._csr.csr_matrix, dtype('int64'))
>>> Y = X * 1.5
>>> type(Y), Y.indptr.dtype
(scipy.sparse._csr.csr_matrix, dtype('int32'))
```

That the values of indices and indptrs were checked to see if they could be coerced to `int32` is a performance problem.

-------------

One more wrinkle to add here. What about `uint`s? `pydata/sparse` and `GraphBLAS` use unsigned integers for their `indices` and `indptr`s. There are upsides and downsides to this.

The most obvious upside being a doubling of addressable coordinates. One use case I've run into here is using a sparse array to store bloom filters.

The biggest downside being numpy's type promotion rules for unsigned integers being unintuitive. Though I thought I heard this might change?

**Comment by rgommers** [comment]
I think there's two separate topics: what's the better design, and what should we actually do given backwards compatibility constraints. 

Regarding the design, I agree with @jjerphan that the current behavior is problematic. I think there''s a by now well-established consensus that it's a bad idea to have output dtype depend on input _values_ rather than input _dtypes_ (for multiple reasons, from code complexity to predictable behavior for users to issues for a JIT compiler like Numba to support a feature).

Regarding backwards compatibility, it's less clear whether we should only add a new option or whether we should try to change the default behavior. I'm not yet convinced one way or the other. Let's see what others think. @perimosocordiae do you have an opinion on this?

> I think the main work to be done is removing steps which automatically downcast the array.

Oof, that does look bad indeed.

> The biggest downside being numpy's type promotion rules for unsigned integers being unintuitive. Though I thought I heard this might change?

Yes, this should improve, but that's not a fast process (see NEP 50).

**Comment by perimosocordiae** [comment]
I'm open to making more broad ranging changes for the new sparse array types, as the need for backward compatibility there is as yet minimal.

For sparse matrices, changing the default index dtype is more difficult. We definitely have users for which 64-bit indices would be a breaking change due to the increased RAM requirement. Removing the down casting behavior should be doable, though.

**Comment by ogauthe** [comment]
I have also been annoyed by the changing index dtype. I would favor switching to `np.int64` as default, since having a fix dtype everywhere is much simpler. In my case, I use numba to construct sparse matrices and `np.int64` is the default int dtype. If the matrix dimension is below `2**31`, one can reasonably expect it to be lighter than a larger one which will need `int64` anyway.

**Comment by rgommers** [comment]
> One more wrinkle to add here. What about `uint`s? `pydata/sparse` and `GraphBLAS` use unsigned integers for their `indices` and `indptr`s. There are upsides and downsides to this.

That does not seem like a good idea to me. Like for dense arrays, the expectation is signed integers for indexing. Aside from it being a large break, people use negative indices like:
```python
>>> row = np.array([0, 0, 1, 2, 2, 2])
... col = np.array([0, 2, 2, 0, 1, 2])
... data = np.array([1, 2, 3, 4, 5, 6])
... x = csr_array((data, (row, col)), shape=(3, 3))
>>> x[0, -1]
2
```

> I think it would be convenient to make to possible to pass an index dtype explicitly at sparse matrix creation time

All the constructors like `csr_matrix` as well as most functions that construct matrices like `sparse.block_diag` already have a `dtype` argument. The functions that are missing it should indeed get that keyword added - that could be a first PR here.

> Removing the down casting behavior should be doable, though.

That could be PR number two.

> I'm open to making more broad ranging changes for the new sparse array types,

That'd be PR number three.

> For sparse matrices, changing the default index dtype is more difficult. We definitely have users for which 64-bit indices would be a breaking change due to the increased RAM requirement. 

Okay, so that seems to be the one hairy one. It's not easy to do a deprecation here, because we don't want to force everyone to use a `dtype=` argument everywhere I'd think. And a global setting doesn't have a precedent in SciPy. So it seems like this should stay as it is, unfortunately.

**Comment by Micky774** [comment]
> All the constructors like `csr_matrix` as well as most functions that construct matrices like `sparse.block_diag` already have a `dtype` argument. The functions that are missing it should indeed get that keyword added - that could be a first PR here.

Regarding that `dtype` argument, isn't that only for the underlying `data` array, not the `index, indptr` arrays? So we would need an orthogonal keyword, e.g. `idx_dtype` as mentioned here: https://github.com/scipy/scipy/pull/16628#discussion_r969998018

**Comment by rgommers** [comment]
> Regarding that `dtype` argument, isn't that only for the underlying `data` array, not the `index, indptr` arrays? So we would need an orthogonal keyword, e.g. `idx_dtype` as mentioned here: [#16628 (comment)](https://github.com/scipy/scipy/pull/16628#discussion_r969998018)

@Micky774 yes, you're correct.

**Comment by jjerphan** [comment]
> > All the constructors like `csr_matrix` as well as most functions that construct matrices like `sparse.block_diag` already have a `dtype` argument. The functions that are missing it should indeed get that keyword added - that could be a first PR here.
> 
> Regarding that `dtype` argument, isn't that only for the underlying `data` array, not the `index, indptr` arrays? So we would need an orthogonal keyword, e.g. `idx_dtype` as mentioned here: [#16628 (comment)](https://github.com/scipy/scipy/pull/16628#discussion_r969998018)

@Micky774: are you interested experimenting with exposing `idx_dtype`?

**Comment by Micky774** [comment]
> @Micky774: are you interested experimenting with exposing `idx_dtype`?

Indeed I am. I'll hopefully start on this soon

**Comment by ivirshup** [comment]
One more point on index dtypes I'd like to bring up, can the `indptr` a `indices` have different dtypes? E.g. `indices_dtype` and `indptr_dtype`?

Where I've run into the need for 64 bit values, it's typically only been on the `indptr`. However, the `indices` array will be quite large, so it would be nice to keep it 32 bit. I imagine there are cases where `intptr` could be kept low bitwidth, while a higher width is needed for `indices`.

**Comment by Micky774** [comment]
> One more point on index dtypes I'd like to bring up, can the `indptr` a `indices` have different dtypes? E.g. `indices_dtype` and `indptr_dtype`?
> 
> Where I've run into the need for 64 bit values, it's typically only been on the `indptr`. However, the `indices` array will be quite large, so it would be nice to keep it 32 bit. I imagine there are cases where `intptr` could be kept low bitwidth, while a higher width is needed for `indices`.

As it stands right now, I don't believe so, since the templated C code regarding these matrices and their operations assumes the same type for both. It shouldn't be impossible to decouple their types, but it would be a pretty comprehensive change since it is coupled throughout a bunch of code iirc.

**Comment by jjerphan** [comment]
> As it stands right now, I don't believe so, since the templated C code regarding these matrices and their operations assumes the same type for both.

Yes, and supporting different types would mean generating a lot of concrete handle functions in thunk. This would in turn makes the sizes of the shared library larger, which was (and still is?) a concern based comments from @perimosocordiae, IIRC.

**Comment by ivirshup** [comment]
Currently working on a solution in: https://github.com/scipy/scipy/pull/18509

I'm starting with what I think is a fairly narrowly scoped idea.

* Never downcast indices types based on values in the indices array
* This is only done for array-api sparse arrays, NOT the matrix classes (e.g. no breaking changes for matrix api classes)

This is not doing anything new to resolve having different dtypes for indices and indptr. So no passing an indices dtype, no new attribute.

> That does not seem like a good idea to me. Like for dense arrays, the expectation is signed integers for indexing. Aside from it being a large break, people use negative indices like:

I've also come around on this. I agree using uints here would likely have unintended consequences.

**Comment by jjerphan** [comment]
Thank you for starting this work, @ivirshup.

I think we need to keep the current behavior identical as the default one (that is preserving downcasting) and provide a way for users to specify the type of indices they want to use.

In terms of UX and UI, might this be specifiable as follows?

```python
from scipy.sparse import csr_matrix

# X.indices and X.indptr will use the smallest bitwidth when possible.
# This would preserve the current behavior (i.e. downcast is still present to save memory
# when possible, and upcast is performed when higher precision is needed). Hence users
# delegate the management of `indices` and `indptr`'s type to SciPy.
X = csr_matrix([1., 2,, 3.])

# X.indices and X.indptr will use the specified type during the whole lifetime of X:
# Users are or can be responsible for the management of `indices` and `indptr`'s type; in this
# case SciPy would not perform any check for representability or potential integer overflow.
X = csr_matrix([1., 2., 3.], indices_dtype=np.int64)
```

What do you think?

**Comment by Micky774** [comment]
> Never downcast indices types based on values in the indices array

I wonder if we could preserve that as existing behavior with e.g. `indices_dtype=None`

It is helpful in many cases I think.

**Comment by perimosocordiae** [comment]
I believe the current idea from @ivirshup is to leave the spmatrix behavior unchanged, and only for sparse arrays update the `get_index_dtype` behavior to avoid checking the index array values and rely solely on the existing index dtype and shape information. This will result in some cases staying in 64-bit land where they technically could fit in 32 bits, but it has some important benefits: we'd no longer perform an expensive scan over the index array(s), and wouldn't trigger a copy either.

**Comment by Micky774** [comment]
I see, I think I had misunderstood the scope of the change. So to cement my understanding, this is meant as a first-step which essentially enforces that sparse arrays "trust `maxval`" when calculating the index dtype right? This is separate from being able to directly specify the dtype, which is what my original comment was (incorrectly) providing feedback for.

**Comment by ivirshup** [comment]
> this is meant as a first-step 

Yes. I have left resolving if we can do different dtypes for different indices arrays and any additional arguments to a later PR.


> enforces that sparse arrays "trust maxval"

One minor clarification here:

Arrays only look at maxval and the current index dtypes (if those are passed). E.g. if you pass in int64 row, col indices for a (3, 3) coo array, the result will have int64 row, col indices.

So really it's "no down casting of passed indices".

You can "specify" the dtype by setting the dtypes of the passed indices.

**Comment by ivirshup** [comment]
The PR is also now ready for review if you want to take a look.

**Comment by ivirshup** [comment]
With https://github.com/scipy/scipy/pull/18509 merged, now we have

```python
from scipy import sparse
import numpy as np

coo = sparse.coo_array(
    (
        np.ones(10),
        (
            np.arange(10, dtype=np.int64),
            np.arange(10, dtype=np.int64)
        )
    )
)

coo.row.dtype
# np.int64
```

🎉

This was accomplished by never downcasting based on the values in the index arrays, so it's also a big performance boost for instantiation of large sparse arrays.

The behaviour for the matrix classes is unchanged.

## Unresolved topics

The PR leaves a few issues discussed here unresolved

### Mixed index dtype support

The PR doesn't change anything about supporting mixd types of index (e.g. having int64 indptr, int32 indices). I would really like to see this (as mentioned above: https://github.com/scipy/scipy/issues/16774#issuecomment-1276486248). I think this can have big performance impacts for:

* Large `csr`, `csc` arrays, where `indptr` needs to be 64 bit due to the number of values stored, but `indices` does not (common in my field)
* Sparse arrays which are very large along one dimension, so only one of `indices`/ `row`/ `col` needs the wider bit type

@jjerphan made a case against this here: https://github.com/scipy/scipy/issues/16774#issuecomment-1277123128

> Yes, and supporting different types would mean generating a lot of concrete handle functions in thunk. This would in turn makes the sizes of the shared library larger, which was (and still is?) a concern based on comments from [@perimosocordiae](https://github.com/perimosocordiae), IIRC.

I don't know how big of an issue this is, and would defer to @perimosocordiae on that. 

### Argument to set index dtype

I didn't add any argument for setting the index dtype. Right now, you "specify" the dtype by setting the dtypes of the passed indices.

Being able to set the index dtype with an argument at construction could be quite useful. Some examples where you can't control the dtype, but maybe you'd want to:

```python
# Using the matrix defined above:
coo.tocsr()

sparse.eye(100)
```

Desired semantics here could intersect with mixed index dtype support.

**Comment by jjerphan** [comment]
Currently, `int32` and `int64` are the only types supported for indices and indices pointers:

https://github.com/scipy/scipy/blob/6c4a3f3551f9ee52a75c8c3999a57bed8ea67deb/scipy/sparse/_generate_sparsetools.py#L136-L141

If we want to support mixed index dtypes with the currently supported index types, this will double the number of concrete functions that will be generated, roughly doubling the size of the shared object (from 3.8MioB to 7.6MioB on Linux). This seems doable.

**Comment by perimosocordiae** [comment]
The situation is slightly worse than that: https://github.com/scipy/scipy/blob/6000091cef20f0025465d2943b0367f546a5a17b/scipy/sparse/sparsetools/csr.h#L691-L696

We have a number of C++ functions that operate on two sparse inputs, so to fully support mixed index dtypes we'd need to generate 16 different versions (all combinations of 32/64 for 4 input arrays) instead of the 2 we currently generate.

Even then I'm not sure if this would push the sparsetools shared library size to unacceptable levels, but it's worth mentioning I think.

**Comment by rgommers** [comment]
> Roughly doubling the size of the shared object (from 3.8MioB to 7.6MioB on Linux).

This would be fairly painful. `scipy.sparse` is already the largest submodule we have, taking (for a manylinux x86-64 wheel):
- 9.3 MB out of 67.6 MB for an installed release build
- 3.5 MB out of 22.2 MB for the `scipy/` contents of a wheel
- 143 MB out of 695 MB for a default development build

In addition, and probably a worse impact, is on build times. The `_sparsetools` C++ files already take by some distance the most build time. `bsr.cxx` alone is the single-file bottleneck for the overall build time of a parallel build (with `csr.cxx` not far behind), see for example the ninjatracing graph at the top of https://labs.quansight.org/blog/2021/07/moving-scipy-to-meson. 

If mixed-integer support would indeed double the build time in addition to the extension module size, as I'd expect, I think that's a no-go. If it's binary sizes alone, I think it's still painful but we can consider it.

> We have a number of C++ functions that operate on two sparse inputs, so to fully support mixed index dtypes we'd need to generate 16 different versions (all combinations of 32/64 for 4 input arrays) instead of the 2 we currently generate.

That would definitely not work.

**Comment by jjerphan** [comment]
#5230 gives examples in favor of mixed index type support.

**Comment by matospiso** [comment]
I agree with #5230. 64-bit indices are often unnecessary and cause large memory overhead.
